src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -149,45 +193,116 @@ internal static SafeSslHandle AllocateSslContext(SslProtocols protocols, SafeX50
+            SafeSslContextHandle? sslCtx = null;
+            SafeSslContextHandle? innerContext = null;
+            SslProtocols protocols = sslAuthenticationOptions.EnabledSslProtocols;
+            bool cacheSslContext = !DisableTlsResume && sslAuthenticationOptions.EncryptionPolicy == EncryptionPolicy.RequireEncryption &&
+                     sslAuthenticationOptions.CipherSuitesPolicy == null &&
+                    (!sslAuthenticationOptions.IsServer ||
+                    (sslAuthenticationOptions.ApplicationProtocols != null && sslAuthenticationOptions.ApplicationProtocols.Count != 0));
+
+            if (!Interop.Ssl.Tls13Supported)
+            {
+                if (protocols != SslProtocols.None &&
+                    CipherSuitesPolicyPal.WantsTls13(protocols))
                 {
-                    unsafe
-                    {
-                        Ssl.SslCtxSetVerify(innerContext, &VerifyClientCertificate);
-                    }
+                    protocols = protocols & (~SslProtocols.Tls13);
+                }
+            }
+            else if (CipherSuitesPolicyPal.WantsTls13(protocols) &&
+                        CipherSuitesPolicyPal.ShouldOptOutOfTls13(sslAuthenticationOptions.CipherSuitesPolicy, sslAuthenticationOptions.EncryptionPolicy))
+            {
+                if (protocols == SslProtocols.None)
+                {
+                    // we are using default settings but cipher suites policy says that TLS 1.3
+                    // is not compatible with our settings (i.e. we requested no encryption or disabled
+                    // all TLS 1.3 cipher suites)
+                    protocols = SslProtocols.Tls | SslProtocols.Tls11 | SslProtocols.Tls12;
+                }
+                else
+                {
+                    // user explicitly asks for TLS 1.3 but their policy is not compatible with TLS 1.3
+                    throw new SslException(
+                        SR.Format(SR.net_ssl_encryptionpolicy_notsupported, sslAuthenticationOptions.EncryptionPolicy));
                 }
+            }
 
+            if (CipherSuitesPolicyPal.ShouldOptOutOfLowerThanTls13(sslAuthenticationOptions.CipherSuitesPolicy, sslAuthenticationOptions.EncryptionPolicy))
+            {
+                if (!CipherSuitesPolicyPal.WantsTls13(protocols))
+                {
+                    // We cannot provide neither TLS 1.3 or non TLS 1.3, user disabled all cipher suites
+                    throw new SslException(
+                        SR.Format(SR.net_ssl_encryptionpolicy_notsupported, sslAuthenticationOptions.EncryptionPolicy));
+                }
+
+                protocols = SslProtocols.Tls13;
+            }
+
+            if (sslAuthenticationOptions.CertificateContext != null && cacheSslContext)
+            {
+               sslAuthenticationOptions.CertificateContext.SslContexts.TryGetValue(sslAuthenticationOptions.EnabledSslProtocols, out sslCtx);
```

Comment:
A lot of work was just done on rewriting the `protocols` local, then this switches back to the authentication options version.

If it's correct that this is not using the local variable, please add a comment explaining why.